docs/src/userguide/language_basics.rst

```diff
@@ -128,75 +128,91 @@ the declaration in most cases:
-                hard = 1
-                soft = 2
-                runny = 3
+                ctypedef int* IntPtr
 
-        There is currently no special syntax for defining a constant, but you can use
-        an anonymous :keyword:`enum` declaration for this purpose, for example,::
+.. _structs:
 
-            cdef enum:
-                tons_of_spam = 3
+Structs, Uions, Enums
+---------------------
 
-        .. note::
-            the words ``struct``, ``union`` and ``enum`` are used only when
-            defining a type, not when referring to it. For example, to declare a variable
-            pointing to a ``Grail`` struct, you would write::
+In addition to the basic types, C :keyword:`struct`, :keyword:`union` and :keyword:`enum`
+are supported:
 
-                cdef Grail *gp
+.. tabs::
 
-            and not::
+    .. group-tab:: Pure Python
 
-                cdef struct Grail *gp  # WRONG
+        .. literalinclude:: ../../examples/userguide/language_basics/struct.py
 
-.. note::
+    .. group-tab:: Cython
 
-    There is also support for giving names to types using the
-    ``ctypedef`` statement or the ``cython.typedef()`` function, e.g.
+        .. literalinclude:: ../../examples/userguide/language_basics/struct.pyx
 
-    .. tabs::
+    .. note::
 
-        .. group-tab:: Pure Python
+        Structs can be declared as ``cdef packed struct``, which has
+        the same effect as the C directive ``#pragma pack(1)``.
 
-            .. code-block:: python
+.. tabs::
 
-                 ULong = cython.typedef(cython.ulong)
+    .. group-tab:: Pure Python
 
-                 IntPtr = cython.typedef(cython.p_int)
+        .. literalinclude:: ../../examples/userguide/language_basics/union.py
 
-        .. group-tab:: Cython
+    .. group-tab:: Cython
 
-            .. code-block:: cython
+        .. literalinclude:: ../../examples/userguide/language_basics/union.pyx
 
-                ctypedef unsigned long ULong
+.. literalinclude:: ../../examples/userguide/language_basics/enum.pyx
 
-                ctypedef int* IntPtr
 
+.. note:: Currently, Pure Python mode does not support enums. (GitHub issue :issue:`4252`)
+
+Declaring an enum as ``cpdef`` will create a :pep:`435`-style Python wrapper::
+
+    cpdef enum CheeseState:
+        hard = 1
+        soft = 2
+        runny = 3
+
+There is currently no special syntax for defining a constant, but you can use
+an anonymous :keyword:`enum` declaration for this purpose, for example,::
+
+    cdef enum:
+        tons_of_spam = 3
+
+.. note::
+    In the Cython syntax, the words ``struct``, ``union`` and ``enum`` are used only when
+    defining a type, not when referring to it. For example, to declare a variable
+    pointing to a ``Grail`` struct, you would write::
+
+        cdef Grail *gp
+
+    and not::
+
+        cdef struct Grail *gp  # WRONG
 
 You can create a C function by declaring it with :keyword:`cdef` or by decorating a Python function with ``@cfunc``:
```

Comment:
Following section/part contains mentions about `cdef` functions and extension types. In general this is wrong place and I think it should be moved:
1. `cdef` function part should be removed because it is better explained in different part (maybe just do double check whether there is not something not mentioned elsewhere.
2. extension type part should be moved somewhere else - not sure where ...

I did not do this changes because I am not 100% sure what is the best. But I am willing to do it after suggestion from others...

Reply:
I think this PR at least needs an extra heading to separate it from the `structs/unions/enums` section. Not thought about the where it should be long-term

Reply:
There are already two sections discussing both extension types and C functions:
* C functions - I have removed this part. There is already full pledged section which mentions everything even better
* Extension types - There is another part mentioning that. I have merged them together and created separated heading for that.